### PR TITLE
nuke: cleanup stale OpenStack resources

### PR DESCRIPTION
**Pull request by dachary**

### CODE REVIEW

**Comment by dachary** [comment]
@zmc the general idea is to replace the scripts you currently have in your script directory.

File: scripts/nuke.py [hunk 1]
```diff
@@ -18,6 +19,7 @@
                         yaml config containing machines to nuke
   -a DIR, --archive DIR
                         archive path for a job to kill and nuke
+  --stale-openstack     nuke 'stale' OpenStack instances
```
Comment (zmc) [inline scripts/nuke.py]: This should go after the ``--stale`` entry and also mention volumes IMO

File: teuthology/nuke.py [hunk 2]
```diff
@@ -361,6 +367,101 @@ def synch_clocks(remotes):
...
+    for volume in volumes:
+        volume = json.loads(sh("openstack volume show -f json " +
+                               volume['ID']))
+        volume = dict(map(lambda v: (v['Field'], v['Value']), volume))
+        created_at = datetime.datetime.strptime(
+            volume['created_at'], '%Y-%m-%dT%H:%M:%S.%f')
+        created = (now - created_at).total_seconds()
+        if created > ctx.teuthology_config['max_job_time'] + OPENSTACK_DELAY:
+            log.info("stale-openstack: volume " + volume['display_name'] +
+                     "(" + volume['id'] + ") destroyed because it"
```
Comment (zmc) [inline teuthology/nuke.py]: Using the paste tense before destruction here will be confusing

File: teuthology/nuke.py [hunk 3]
```diff
@@ -361,6 +367,101 @@ def synch_clocks(remotes):
...
+                     " because it was created " + str(i.get_created()) +
+                     " seconds ago which is older than max_job_time " +
+                     str(ctx.teuthology_config['max_job_time']) + " + " +
+                     str(OPENSTACK_DELAY))
+            if not ctx.dry_run:
+                i.destroy()
+            continue
+        name = canonicalize_hostname(i['name'], user=None)
+        if i.get_created() > OPENSTACK_DELAY and name not in locked_nodes:
+            log.info("stale-openstack: instance " + i['name'] + " destroyed" +
```
Comment (zmc) [inline teuthology/nuke.py]: Using the paste tense before destruction here will be confusing

File: teuthology/nuke.py [hunk 4]
```diff
@@ -361,6 +367,101 @@ def synch_clocks(remotes):
...
+        log.debug("stale-openstack: volume " + volume['id'] + " OK")
+
+def stale_openstack_nodes(ctx, instances, locked_nodes):
+    for (name, node) in locked_nodes.iteritems():
+        name = decanonicalize_hostname(name)
+        if node['machine_type'] != 'openstack':
+            continue
+        if (name not in instances and
+            locked_since_seconds(node) > OPENSTACK_DELAY):
+            log.info("stale-openstack: node " + name + " unlocked" +
```
Comment (zmc) [inline teuthology/nuke.py]: Using the paste tense before unlocking here will be confusing

File: teuthology/nuke.py [hunk 5]
```diff
@@ -361,6 +367,101 @@ def synch_clocks(remotes):
...
+                     str(OPENSTACK_DELAY) + " and it has no instance")
+            if not ctx.dry_run:
+                unlock_one(ctx, name, node['locked_by'])
+            continue
+        log.debug("stale-openstack: node " + name + " OK")
+
+def openstack_remove_again():
+    """Volumes and servers with REMOVE-ME in the name are leftover
+    that failed to be removed. It is not uncommon for a failed removal
+    to succeed later on."""
```
Comment (zmc) [inline teuthology/nuke.py]: Please make this docstring look:
```
"""
like this
"""

File: teuthology/nuke.py [hunk 6]
```diff
@@ -361,6 +367,101 @@ def synch_clocks(remotes):
...
+    xargs --no-run-if-empty --max-args 1 -P20 openstack server delete --wait
+    true
+    """)
+    sh("""
+    openstack volume list --name REMOVE-ME --column ID --format value |
+    xargs --no-run-if-empty --max-args 1 -P20 openstack volume delete
+    true
+    """)
+
+def stale_openstack(ctx):
```
Comment (zmc) [inline teuthology/nuke.py]: Could this be moved above the other functions it calls? It'll help with future readability

File: teuthology/nuke.py [hunk 7]
```diff
@@ -361,6 +367,101 @@ def synch_clocks(remotes):
...
+# any kind of OpenStack server creation / deletion / etc.
+#
+OPENSTACK_DELAY = 30 * 60
+
+def stale_openstack_instances(ctx, instances, locked_nodes):
+    for (name, instance) in instances.iteritems():
+        i = OpenStackInstance(name)
+        if (i.get_created() >
+            ctx.teuthology_config['max_job_time'] + OPENSTACK_DELAY):
+            log.info("stale-openstack: instance " + i['name'] + " destroyed" +
```
Comment (zmc) [inline teuthology/nuke.py]: More paste tense before destruction. Seems like these strings would benefit from templating either using the ``%s``-style replacement or ``str.format()`` - then, this message could be reused and extended below optionally

File: teuthology/openstack/__init__.py [hunk 8]
```diff
@@ -42,6 +43,111 @@
...
+    def get_created(self):
+        now = datetime.datetime.now()
+        created = datetime.datetime.strptime(
+            self['created'], '%Y-%m-%dT%H:%M:%SZ')
+        return (now - created).total_seconds()
+
+    def exists(self):
+        return self.info is not None
+
+    def volumes(self):
```
Comment (zmc) [inline teuthology/openstack/__init__.py]: Should this be ``get_volumes()`` to follow the class's existing convention?

File: scripts/nuke.py (hunk 1 above)
Comment (dachary) [inline scripts/nuke.py]: done

File: teuthology/nuke.py (hunk 3 above)
Comment (dachary) [inline teuthology/nuke.py]: done

File: teuthology/nuke.py (hunk 2 above)
Comment (dachary) [inline teuthology/nuke.py]: done

File: teuthology/nuke.py (hunk 4 above)
Comment (dachary) [inline teuthology/nuke.py]: done

File: teuthology/nuke.py (hunk 5 above)
Comment (dachary) [inline teuthology/nuke.py]: done

File: teuthology/nuke.py (hunk 2 above)
Comment (dachary) [inline teuthology/nuke.py]: done

File: teuthology/nuke.py (hunk 6 above)
Comment (dachary) [inline teuthology/nuke.py]: done

File: teuthology/nuke.py (hunk 3 above)
Comment (dachary) [inline teuthology/nuke.py]: done

File: teuthology/nuke.py (hunk 7 above)
Comment (dachary) [inline teuthology/nuke.py]: done (no factorization but all converted to format())

File: teuthology/openstack/__init__.py (hunk 8 above)
Comment (dachary) [inline teuthology/openstack/__init__.py]: done

**Comment by dachary** [comment]
@zmc repushed with the suggested changes :-)

**Pull request merged**